tests/components/geo_location/test_geo_json_events.py

```diff
@@ -27,200 +24,231 @@
+            async_fire_time_changed(hass, utcnow + 2 * SCAN_INTERVAL)
+            await hass.async_block_till_done()
+
+            all_states = hass.states.async_all()
+            assert len(all_states) == 3
+
+            # Simulate an update - empty data, removes all entities
+            mock_feed.return_value.update.return_value = 'ERROR', None
+            async_fire_time_changed(hass, utcnow + 3 * SCAN_INTERVAL)
+            await hass.async_block_till_done()
+
+            all_states = hass.states.async_all()
+            assert len(all_states) == 0
+
+
+async def test_setup_with_custom_location(hass):
+    """Test the setup with a custom location."""
+    # Set up some mock feed entries for this test.
+    mock_entry_1 = _generate_mock_feed_entry(
+        '1234', 'Title 1', 2000.5, (-31.1, 150.1))
+
+    with patch('geojson_client.generic_feed.GenericFeed') as mock_feed:
+        mock_feed.return_value.update.return_value = 'OK', [mock_entry_1]
+
+        with assert_setup_component(1, geo_location.DOMAIN):
+            assert await async_setup_component(
+                hass, geo_location.DOMAIN, CONFIG_WITH_CUSTOM_LOCATION)
+
+            # Artificially trigger update.
+            hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
+            # Collect events.
+            await hass.async_block_till_done()
+
+            all_states = hass.states.async_all()
+            assert len(all_states) == 1
+
+            assert mock_feed.call_args == call(
+                (15.1, 25.2), URL, filter_radius=200.0)
+
+
+async def test_setup_race_condition(hass):
+    """Test a particular race condition experienced."""
+    # 1. Feed returns 1 entry -> Feed manager creates 1 entity.
+    # 2. Feed returns error -> Feed manager removes 1 entity.
+    #    However, this stayed on and kept listening for dispatcher signals.
+    # 3. Feed returns 1 entry -> Feed manager creates 1 entity.
+    # 4. Feed returns 1 entry -> Feed manager updates 1 entity.
+    #    Internally, the previous entity is updating itself, too.
+    # 5. Feed returns error -> Feed manager removes 1 entity.
+    #    There are now 2 entities trying to remove themselves from HA, but
+    #    the second attempt fails of course.
+
+    # Set up some mock feed entries for this test.
+    mock_entry_1 = _generate_mock_feed_entry(
+        '1234', 'Title 1', 15.5, (-31.0, 150.0))
+
+    # Patching 'utcnow' to gain more control over the timed update.
+    utcnow = dt_util.utcnow()
+    with patch('homeassistant.util.dt.utcnow', return_value=utcnow), \
+            patch('geojson_client.generic_feed.GenericFeed') as mock_feed:
+        with assert_setup_component(1, geo_location.DOMAIN):
+            assert await async_setup_component(
+                hass, geo_location.DOMAIN, CONFIG)
+
+            # This gives us the ability to assert the '_delete_callback'
+            # has been called while still executing it.
+            original_delete_callback = homeassistant.components\
```

Comment:
We don't need to assert that this callback has been called directly. Just check that the state of the entity has been removed from the state machine.

Reply:
I understand. The purpose of this particular test case was to ensure that the `_delete_callback` was guaranteed to be only called once when an entity is removed. Without the bugfix, the `_delete_callback` was called once, the entity was removed, and then `_delete_callback` was called again leading to an error.
If you like I can as well remove this test case now after the bug has been fixed. The fix was to keep a handle of the dispatcher connects and do the callback when the entity is removed.

Reply:
Can't remember, but I can easily reproduce this issue.

Reply:
OK, reproduced the original issue:
There is no `Error` raised while running Home Assistant, but what basically happens is that if an entity object is successfully removed from the state machine, the entity object lives on in memory and remains attached to the dispatcher and continues to process updates. Over time, with a high turnover external feed, more and more orphaned entities remain in memory and use up valuable processing time.

Reply:
Ok. The only other approach for testing this that I can think of is to check the dispatchers signal list in hass.data. But I'm not sure this is better.
```py
hass.data[DATA_DISPATCHER][signal]
```

You decide.

The reason why I'd like to avoid asserting things about the entity, is that it makes the test less robust, since the entity implementation might change.

Reply:
Thanks. I like this approach much better - even though it looks into the internal state of the dispatcher, it's easier to set up than mocking a protected method, and hopefully easier to understand.

What I just noticed while debugging was that the dict in `hass.data[DATA_DISPATCHER]` keeps its keys (=signals) even after all targets have been removed. The `geo_location` platforms now all use unique signals that contain the external feed entry's unique id, i.e. over time this dict will contain many keys without targets. Just wondering if the dispatcher may need some sort of clean-up method in the long term?

Reply:
Yeah, that's a good point. Ask in discord dev or make an architecture issue.